### PR TITLE
Talk about concurrency in H3 vs. H2

### LINKED ISSUE #2787: HTTP/3 has less concurrency than HTTP/2
At the moment, we state that "It is RECOMMENDED that initial_max_bidi_streams be no smaller than 100, so as to not unnecessarily limit parallelism."

This means that there is a reduction in parallelism compared to HTTP/2 that allows 100 concurrent requests, due to the following reasons:
1. unlike HTTP/2, a QUIC server is likely to consider a stream to be existing until receiving an ACK for the stream-level FIN that it has sent
2. even after the stream state is discarded (when receiving an ACK for FIN), a server might not immediately send a MAX_STREAMS frame to indicate the client can open new streams

IIUC, one of the primary reasons we want parallelism is to fulfill the bandwidth of a high latency connection even when the size of the HTTP responses are small<sup>2</sup>.

The differences pointed out above means that in such a scenario (where the responses are small, e.g. 304), the concurrency could go down from 100 requests per RTT in HTTP/2 to 50 in HTTP/3, because the minimal latency of retiring a stream state changes from 1 RTT to 2 RTT (see point 1 above).

I think we should provide some better guidance, assuming that we consider "100" that we have in HTTP/2 as a good advice.

One way would be to change the recommended minimum of initial_max_bidi_streams to 200. The other way would be to recommend servers providing clients "window to issue 100 requests in parallel," without giving them concrete suggestion on how, because the way the QUIC servers retain stream-level states and how they send MAX_STREAMS can be different.

2: The other reason is to provide sufficient amount of window for prioritization.

### PR DESCRIPTION
**Pull request by MikeBishop**
Fixes #2787.

### CODE REVIEW

File: draft-ietf-quic-http.md [hunk 1]
```diff
@@ -1814,6 +1814,15 @@ considerations about exhaustion of stream identifier space apply, though the
 space is significantly larger such that it is likely that other limits in QUIC
 are reached first, such as the limit on the connection flow control window.
 
+QUIC manages stream concurrency on behalf of HTTP/3.  QUIC considers a stream
```
Comment (kazuho) [inline draft-ietf-quic-http.md]: I feel this "on behalf of HTTP/3" a bit confusing, because my mental model is that it is generally a responsibility of the transport to manage streams, and that HTTP/2 was the exception.

Maybe something like "In contrast to HTTP/2, stream concurrency in HTTP/3 is managed by QUIC."

File: draft-ietf-quic-http.md [hunk 2]
```diff
@@ -1814,6 +1814,15 @@ considerations about exhaustion of stream identifier space apply, though the
...
+closed when all sent data and all acknowledgements for received data have been
+acknowledged by the peer.  HTTP/2 considers a stream closed when the frame
```
Comment (kazuho) [inline draft-ietf-quic-http.md]: I think this sentence is incorrect.

I think it should be: "QUIC considers a stream closed when all data have been received and all the sent data have been acknowledged."

File: draft-ietf-quic-http.md [hunk 3]
```diff
@@ -1814,6 +1814,15 @@ considerations about exhaustion of stream identifier space apply, though the
...
+containing the END_STREAM bit has been committed to the transport. As a result,
+the stream for an equivalent exchange will remain "active" for at least one
+additional round trip.  Implementations of HTTP/3 might choose to permit a
```
Comment (kazuho) [inline draft-ietf-quic-http.md]: How about "will typically remain "active" for one additional roundtrip"?

The lifetime of a stream would be indifferent from HTTP/2 in the pathological case where the request-side of the stream is closed after the response-side is being closed.

File: draft-ietf-quic-http.md [hunk 4]
```diff
@@ -1814,6 +1814,22 @@ considerations about exhaustion of stream identifier space apply, though the
...
+containing the END_STREAM bit has been committed to the transport. As a result,
+the stream for an equivalent exchange will typically remain "active" for one
+additional round trip.  HTTP/3 servers might choose to permit a larger number of
+concurrent client-initiated bidirectional streams to achieve equivalent
+concurrency than were permitted in HTTP/2, depending on the expected usage
+patterns.
+
+Due to the presence of other unidirectional stream types, HTTP/3 does not rely
+exclusively on the number of concurrent unidirectional streams to control the
+number of concurrent pushes received by a client.  Instead, HTTP/3 clients use
```
Comment (LPardue) [inline draft-ietf-quic-http.md]:
```suggestion
number of concurrent in-flight pushes.  Instead, HTTP/3 clients use
```

Avoid the rabbit whole of received by client or sent by server.

File: draft-ietf-quic-http.md [hunk 5]
```diff
@@ -1814,6 +1814,22 @@ considerations about exhaustion of stream identifier space apply, though the
...
 are reached first, such as the limit on the connection flow control window.
 
+In contrast to HTTP/2, stream concurrency in HTTP/3 is managed by QUIC.  QUIC
+considers a stream closed when all data has been received and sent data has been
+acknowledged by the peer.  HTTP/2 considers a stream closed when the frame
+containing the END_STREAM bit has been committed to the transport. As a result,
+the stream for an equivalent exchange will typically remain "active" for one
+additional round trip.  HTTP/3 servers might choose to permit a larger number of
+concurrent client-initiated bidirectional streams to achieve equivalent
+concurrency than were permitted in HTTP/2, depending on the expected usage
```
Comment (afrind) [inline draft-ietf-quic-http.md]: "permit a larger number...than were permitted" makes sense, but there are so many words between them that it's kind of hard to parse.

File: draft-ietf-quic-http.md (hunk 2 above)
Comment (ianswett) [reply draft-ietf-quic-http.md]: Or a reset has been sent and acknowledged.

File: draft-ietf-quic-http.md (hunk 5 above)
Comment (ianswett) [inline draft-ietf-quic-http.md]:
```suggestion
concurrency to in HTTP/2, depending on the expected usage
```
